### PR TITLE
feat: store sort_key in the chunk schema after the chunk is sorted

### PR DESCRIPTION
**Pull request by NGA-TRAN**
Closes #

This is the first part of #1886 that store the sort key of the chunk after it is compacted and sorted. Next step will use the keys to avoid sort if appropriate

- [ ] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [ ] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: query/src/provider.rs [hunk 1]
```diff
@@ -678,13 +678,16 @@ impl<C: QueryChunk + 'static> Deduplicater<C> {
         chunk: Arc<C>,
         input: Arc<dyn ExecutionPlan>,
     ) -> Result<Arc<dyn ExecutionPlan>> {
+
+        // Todo: check there is sort key and it matches with the given one
+        //let sort_key = schema.sort_key();
```
Comment (NGA-TRAN) [inline query/src/provider.rs:683]: This is my next step after the mutex issue is fixed

File: server/src/db/catalog/chunk.rs [hunk 2]
```diff
@@ -132,7 +133,7 @@ pub enum ChunkStage {
     /// or more chunks and creates a single new frozen chunk.
     Frozen {
         /// Metadata (statistics, schema) about this chunk
-        meta: Arc<ChunkMetadata>,
+        meta: Arc<Mutex<ChunkMetadata>>,
```
Comment (NGA-TRAN) [inline server/src/db/catalog/chunk.rs:136]: Since the chunk is sorted when moving from MUB to RUB, I need to update the schema with the sort key, and need to make this a Mutex

File: server/src/db/chunk.rs [hunk 3]
```diff
@@ -444,10 +449,12 @@ impl QueryChunk for DbChunk {
 
 impl QueryChunkMeta for DbChunk {
     fn summary(&self) -> &TableSummary {
-        self.meta.table_summary.as_ref()
+        self.meta.table_summary.as_ref() // original that is no longer working because of the mutex
+        // have tried may other things but if I do not modify the return value as a MutexGaurd, they won't work. But since this is a function of a trait, I do not want to modify it
```
Comment (NGA-TRAN) [inline server/src/db/chunk.rs:453]: Got get compile error here and need help

File: server/src/db/catalog/chunk.rs [hunk 4]
```diff
@@ -610,30 +617,35 @@ impl CatalogChunk {
...
+    pub fn set_moved(&mut self, chunk: Arc<RBChunk>, schema: Schema) -> Result<()> {
         match &mut self.stage {
-            ChunkStage::Frozen { representation, .. } => match &representation {
-                ChunkStageFrozenRepr::MutableBufferSnapshot(_) => {
-                    self.metrics
-                        .state
-                        .inc_with_labels(&[KeyValue::new("state", "moved")]);
+            ChunkStage::Frozen {meta,  representation, .. } => {
+                // after moved, the chunk is sorted and schema need to get updated
+                let mut metadata = meta.lock();
```
Comment (alamb) [inline server/src/db/catalog/chunk.rs:624]: If this is the only use of the mutex, I think you can avoid using it with something like:

```suggestion
            ChunkStage::Frozen {mut metadata,  representation, .. } => {
                // after moved, the chunk is sorted and schema need to get updated
```

Comment (alamb) [inline server/src/db/catalog/chunk.rs:624]:
```suggestion
            ChunkStage::Frozen {mut metadata,  representation, .. } => {
                // after moved, the chunk is sorted and schema need to get updated
```
Can you do this and avoid the use of the Mutex?
Can you use this for

Comment (NGA-TRAN) [reply server/src/db/catalog/chunk.rs:624]: Let me try again.I thought I did it and it did not work but I might have tried something different

**Comment by mkmik** [comment]
If you need to return a MutexGuard from one of the impls but you don't want to force it to be part of the trait signature, consider using associated types.

Here's a simplified scenario that shows how to define a QueryChunkMeta trait that can return either a `&TableSummary` or a `MutexGuard<TableSummary>`:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=dc7ce555da0d7068f3b0ed7f1d36eb70

File: query_tests/src/scenarios.rs [hunk 5]
```diff
@@ -185,6 +185,40 @@ impl DbSetup for NoData {
 
 /// Two measurements data in a single mutable buffer chunk
 #[derive(Debug)]
+pub struct TwoMeasurementsMubScenario {}
```
Comment (NGA-TRAN) [inline query_tests/src/scenarios.rs:188]: Since now the schema of RUB includes sort key, we cannot not use the same `expected` for both MUB and RUB. I need to split the test into schema for MUB and for RUB

File: internal_types/src/schema.rs [hunk 6]
```diff
@@ -218,6 +218,39 @@ impl Schema {
         Ok(record)
     }
 
+    /// Set the order of sort columns to the primary key id the schema
```
Comment (alamb) [inline internal_types/src/schema.rs:221]:
```suggestion
    /// Set the order of sort columns to the specified `sort_key`
```

File: internal_types/src/schema.rs [hunk 7]
```diff
@@ -218,6 +218,39 @@ impl Schema {
...
+        let new_fields = fields
+            .iter()
+            .map(|field| {
+                let mut new_field = field.clone();
+                if let Some(sort) = sort_key.get(field.name()) {
+                    let mut meta = std::collections::BTreeMap::new();
+                    // New sort key
+                    meta.insert(COLUMN_SORT_METADATA_KEY.to_string(), sort.to_string());
+                    // Keep other meta data
+                    if let Some(metadata) = field.metadata().to_owned() {
```
Comment (alamb) [inline internal_types/src/schema.rs:235]: I wonder if you have to do `to_owned()` here (which makes a copy) -- can you instead do something like

```suggestion
                    if let Some(metadata) = field.metadata() {
```

File: server/src/db/chunk.rs [hunk 8]
```diff
@@ -432,7 +432,11 @@ impl QueryChunk for DbChunk {
         }
     }
 
-    // TODOs: return the right value. For now the chunk is assumed to be not sorted
+    /// Returns true if the chunk is sorted on its pk
+    /// Since data is compacted prior being moved to RUBs, data in RUBs and OBs
```
Comment (alamb) [inline server/src/db/chunk.rs:436]: I am not sure what a `OBs` means

**Review by alamb — APPROVED** [review]
The only thing I think that would make this PR better would be some unit tests for `Schema::set_sort_key` --   Specifically, the cases I was thinking about are:

1. What happens if the sort key refers to columns that are not in the schema (I think they are ignored now, but maybe that should be an error)
2. What happens if there was an existing sort key `A, B, C` and then a new sort key is set `A, B` -- does the resulting schema only have `A, B` as sort keys? I think it might still be `A, B, C`

File: query/src/frontend/reorg.rs [hunk 9]
```diff
@@ -73,10 +73,14 @@ impl ReorgPlanner {
...
             plan_builder,
             provider,
-        } = self.scan_and_sort_plan(chunks, output_sort)?;
+        } = self.scan_and_sort_plan(chunks, output_sort.clone())?;
+
+        let mut schema = provider.iox_schema();
 
-        // TODO: Set sort key on schema
-        let schema = provider.iox_schema();
+        trace!("Schema before setting sort key: {:#?}", schema);
```
Comment (e-dard) [inline query/src/frontend/reorg.rs:80]: When logging it is preferable to keep the log message fixed and then provide structured keys and values. This will help when processing logs or grepping for specific log events.

```suggestion
        trace!(input_schema=?schema, "Setting sort key on schema");
```

File: query/src/frontend/reorg.rs [hunk 10]
```diff
@@ -73,10 +73,14 @@ impl ReorgPlanner {
...
+        // Set the sort_key of the schema to the compacted chunk's sort key
+        schema.set_sort_key(&output_sort);
+        trace!("Schema after setting sort key: {:#?}", schema);
```
Comment (e-dard) [inline query/src/frontend/reorg.rs:83]:
```suggestion
        trace!(output_schema=?schema, "Setting sort key on schema");
```

File: server/src/db/chunk.rs (hunk 8 above)
Comment (e-dard) [reply server/src/db/chunk.rs:436]: I think @NGA-TRAN means Object Stores?

```suggestion
    /// Since data is compacted prior being moved to RUBs, data in RUBs and object store
```

**Review by e-dard — APPROVED** [review]
Looking good.

File: internal_types/src/schema.rs [hunk 11]
```diff
@@ -218,6 +218,39 @@ impl Schema {
...
+                        metadata.iter().for_each(|(key, value)| {
```
Comment (e-dard) [inline internal_types/src/schema.rs:236]: I think it is more idiomatic to use a `for` loop here. `for_each` is probably preferable if you are at the end of a big chain of iterators or if you care about bounds checking, but in this case

```rust
for (key, value) in &metadata {
  ...
}
```

would be preferable imo.

File: internal_types/src/schema.rs (hunk 7 above)
Comment (NGA-TRAN) [reply internal_types/src/schema.rs:235]: Good catch. The to_owned definitely no need here.The original copy intention for different purpose but I did not use it and forgot to remove the to_owned

File: query/src/frontend/reorg.rs (hunk 9 above)
Comment (NGA-TRAN) [reply query/src/frontend/reorg.rs:80]: @e-dard Is this equivalent to your suggestion?
trace!("Setting sort key on schema - input_schema:{:#?}", schema);
I want to have the structure that is easy to read and the `?schema` does not provide that

Comment (e-dard) [reply query/src/frontend/reorg.rs:80]: It's not quite the same. The idea of using structured fields in your logs is that they comply to the logfmt spec, which means any parser can split out the fields into separate values. If we were feeding these log lines into IOx for example we would be able to put them in different columns. 

The message itself is one field and the input schema would be another.

Comment (NGA-TRAN) [reply query/src/frontend/reorg.rs:80]: Let me keep it your format, then. Thanks @e-dard
